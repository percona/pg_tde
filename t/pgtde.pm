package PGTDE;

use PostgreSQL::Test::Cluster;
use PostgreSQL::Test::Utils;

use File::Basename;
use File::Compare;
use Test::More;
use Time::HiRes qw(usleep);

# Expected .out filename of TAP testcase being executed. These are already part of repo under t/expected/*.
our $expected_filename_with_path;

# Result .out filename of TAP testcase being executed. Where needed, a new *.out will be created for each TAP test.
our $out_filename_with_path;

# Runtime output file that is used only for debugging purposes for comparison to PGSS, blocks and timings.
our $debug_out_filename_with_path;

my $expected_folder = "t/expected";
my $results_folder = "t/results";

sub psql
{
	my ($node, $dbname, $sql) = @_;

	my (undef, $stdout, $stderr) = $node->psql($dbname, $sql,
		extra_params => [ '-a', '-Pformat=aligned', '-Ptuples_only=off' ]);

	if ($stdout ne '')
	{
		append_to_result_file($stdout);
	}

	if ($stderr ne '')
	{
		append_to_result_file($stderr);
	}
}

# Copied from src/test/recovery/t/017_shm.pl
sub poll_start
{
	my ($node) = @_;

	my $max_attempts = 10 * $PostgreSQL::Test::Utils::timeout_default;
	my $attempts = 0;

	while ($attempts < $max_attempts)
	{
		$node->start(fail_ok => 1) && return 1;

		# Wait 0.1 second before retrying.
		usleep(100_000);

		# Clean up in case the start attempt just timed out or some such.
		$node->stop('fast', fail_ok => 1);

		$attempts++;
	}

	# Try one last time without fail_ok, which will BAIL_OUT unless it
	# succeeds.
	$node->start && return 1;
	return 0;
}

sub append_to_result_file
{
	my ($str) = @_;

	append_to_file($out_filename_with_path, $str . "\n");
}

sub append_to_debug_file
{
	my ($str) = @_;

	append_to_file($debug_out_filename_with_path, $str . "\n");
}

sub setup_files_dir
{
	my ($test_filename) = @_;

	unless (-d $results_folder)
	{
		mkdir $results_folder
		  or die "Can't create folder $results_folder: $!\n";
	}

	my ($test_name) = $test_filename =~ /([^.]*)/;

	$expected_filename_with_path = "${expected_folder}/${test_name}.out";
	$out_filename_with_path = "${results_folder}/${test_name}.out";
	$debug_out_filename_with_path =
	  "${results_folder}/${test_name}.out.debug";

	if (-f $out_filename_with_path)
	{
		unlink($out_filename_with_path)
		  or die
		  "Can't delete already existing $out_filename_with_path: $!\n";
	}
}

sub compare_results
{
	return compare($expected_filename_with_path, $out_filename_with_path);
}

sub backup
{
	my ($node, $backup_name, %params) = @_;
	my $backup_dir = $node->backup_dir . '/' . $backup_name;
	my $name = $node->name;

	local %ENV = $node->_get_env();

	mkdir $backup_dir or die "mkdir($backup_dir) failed: $!";

	PostgreSQL::Test::RecursiveCopy::copypath($node->data_dir . '/pg_tde',
		$backup_dir . '/pg_tde');

	print "# Taking pg_basebackup $backup_name from node \"$name\"\n";
	PostgreSQL::Test::Utils::system_or_bail(
		'pg_tde_basebackup', '-D',
		$backup_dir, '-h',
		$node->host, '-p',
		$node->port, '--checkpoint',
		'fast', '--no-sync',
		'-E', @{ $params{backup_options} });
	print "# Backup finished\n";
	return;
}

1;
