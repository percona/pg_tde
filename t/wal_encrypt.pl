#!/usr/bin/perl

use strict;
use warnings;
use File::Basename;
use Test::More;
use lib 't';
use pgtde;

PGTDE::setup_files_dir(basename($0));

unlink('/tmp/wal_encrypt.per');

my $node = PostgreSQL::Test::Cluster->new('main');
$node->init;
$node->append_conf('postgresql.conf', "shared_preload_libraries = 'pg_tde'");
$node->append_conf('postgresql.conf', "wal_level = 'logical'");
# We don't test that it can't start: the test framework doesn't have an easy way to do this
#$node->append_conf('postgresql.conf', "pg_tde.wal_encrypt = 1");
$node->start;

PGTDE::psql($node, 'postgres', "CREATE EXTENSION pg_tde;");

PGTDE::psql($node, 'postgres',
	"SELECT pg_tde_add_global_key_provider_file('file-keyring-010', '/tmp/wal_encrypt.per');"
);

PGTDE::psql($node, 'postgres', 'SELECT pg_tde_verify_server_key();');

PGTDE::psql($node, 'postgres',
	'SELECT key_name, provider_name, provider_id FROM pg_tde_server_key_info();'
);

PGTDE::psql($node, 'postgres',
	"SELECT pg_tde_create_key_using_global_key_provider('server-key', 'file-keyring-010');"
);
PGTDE::psql($node, 'postgres',
	"SELECT pg_tde_set_server_key_using_global_key_provider('server-key', 'file-keyring-010');"
);

PGTDE::psql($node, 'postgres', 'SELECT pg_tde_verify_server_key();');

PGTDE::psql($node, 'postgres',
	'SELECT key_name, provider_name, provider_id FROM pg_tde_server_key_info();'
);

PGTDE::psql($node, 'postgres', 'ALTER SYSTEM SET pg_tde.wal_encrypt = on;');

PGTDE::append_to_result_file("-- server restart with wal encryption");
$node->restart;

PGTDE::psql($node, 'postgres', "SHOW pg_tde.wal_encrypt;");

PGTDE::psql($node, 'postgres',
	"SELECT slot_name FROM pg_create_logical_replication_slot('tde_slot', 'test_decoding');"
);

PGTDE::psql($node, 'postgres',
	'CREATE TABLE test_wal (id SERIAL, k INTEGER, PRIMARY KEY (id));');

PGTDE::psql($node, 'postgres', 'INSERT INTO test_wal (k) VALUES (1), (2);');

PGTDE::psql($node, 'postgres', 'ALTER SYSTEM SET pg_tde.wal_encrypt = off;');

PGTDE::append_to_result_file("-- server restart without wal encryption");
$node->restart;

PGTDE::psql($node, 'postgres', "SHOW pg_tde.wal_encrypt;");

PGTDE::psql($node, 'postgres', 'INSERT INTO test_wal (k) VALUES (3), (4);');

PGTDE::psql($node, 'postgres', 'ALTER SYSTEM SET pg_tde.wal_encrypt = on;');

PGTDE::append_to_result_file("-- server restart with wal encryption");
$node->restart;

PGTDE::psql($node, 'postgres', "SHOW pg_tde.wal_encrypt;");

PGTDE::psql($node, 'postgres', 'INSERT INTO test_wal (k) VALUES (5), (6);');

PGTDE::append_to_result_file("-- server restart with still wal encryption");
$node->restart;

PGTDE::psql($node, 'postgres', "SHOW pg_tde.wal_encrypt;");

PGTDE::psql($node, 'postgres', 'INSERT INTO test_wal (k) VALUES (7), (8);');

PGTDE::psql($node, 'postgres',
	"SELECT data FROM pg_logical_slot_get_changes('tde_slot', NULL, NULL, 'include-xids', '0');"
);

PGTDE::psql($node, 'postgres',
	"SELECT pg_drop_replication_slot('tde_slot');");

PGTDE::psql($node, 'postgres', 'DROP EXTENSION pg_tde;');

$node->stop;

# Compare the expected and out file
my $compare = PGTDE->compare_results();

is($compare, 0,
	"Compare Files: $PGTDE::expected_filename_with_path and $PGTDE::out_filename_with_path files."
);

done_testing();
