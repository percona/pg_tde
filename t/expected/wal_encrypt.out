CREATE EXTENSION pg_tde;
SELECT pg_tde_add_global_key_provider_file('file-keyring-010', '/tmp/wal_encrypt.per');
 pg_tde_add_global_key_provider_file 
-------------------------------------
 
(1 row)

SELECT pg_tde_verify_server_key();
psql:<stdin>:1: ERROR:  principal key not configured for current database
SELECT key_name, provider_name, provider_id FROM pg_tde_server_key_info();
 key_name | provider_name | provider_id 
----------+---------------+-------------
          |               |            
(1 row)

SELECT pg_tde_create_key_using_global_key_provider('server-key', 'file-keyring-010');
 pg_tde_create_key_using_global_key_provider 
---------------------------------------------
 
(1 row)

SELECT pg_tde_set_server_key_using_global_key_provider('server-key', 'file-keyring-010');
 pg_tde_set_server_key_using_global_key_provider 
-------------------------------------------------
 
(1 row)

SELECT pg_tde_verify_server_key();
 pg_tde_verify_server_key 
--------------------------
 
(1 row)

SELECT key_name, provider_name, provider_id FROM pg_tde_server_key_info();
  key_name  |  provider_name   | provider_id 
------------+------------------+-------------
 server-key | file-keyring-010 |          -1
(1 row)

ALTER SYSTEM SET pg_tde.wal_encrypt = on;
-- server restart with wal encryption
SHOW pg_tde.wal_encrypt;
 pg_tde.wal_encrypt 
--------------------
 on
(1 row)

SELECT slot_name FROM pg_create_logical_replication_slot('tde_slot', 'test_decoding');
 slot_name 
-----------
 tde_slot
(1 row)

CREATE TABLE test_wal (id SERIAL, k INTEGER, PRIMARY KEY (id));
INSERT INTO test_wal (k) VALUES (1), (2);
ALTER SYSTEM SET pg_tde.wal_encrypt = off;
-- server restart without wal encryption
SHOW pg_tde.wal_encrypt;
 pg_tde.wal_encrypt 
--------------------
 off
(1 row)

INSERT INTO test_wal (k) VALUES (3), (4);
ALTER SYSTEM SET pg_tde.wal_encrypt = on;
-- server restart with wal encryption
SHOW pg_tde.wal_encrypt;
 pg_tde.wal_encrypt 
--------------------
 on
(1 row)

INSERT INTO test_wal (k) VALUES (5), (6);
-- server restart with still wal encryption
SHOW pg_tde.wal_encrypt;
 pg_tde.wal_encrypt 
--------------------
 on
(1 row)

INSERT INTO test_wal (k) VALUES (7), (8);
SELECT data FROM pg_logical_slot_get_changes('tde_slot', NULL, NULL, 'include-xids', '0');
                           data                            
-----------------------------------------------------------
 BEGIN
 COMMIT
 BEGIN
 table public.test_wal: INSERT: id[integer]:1 k[integer]:1
 table public.test_wal: INSERT: id[integer]:2 k[integer]:2
 COMMIT
 BEGIN
 table public.test_wal: INSERT: id[integer]:3 k[integer]:3
 table public.test_wal: INSERT: id[integer]:4 k[integer]:4
 COMMIT
 BEGIN
 table public.test_wal: INSERT: id[integer]:5 k[integer]:5
 table public.test_wal: INSERT: id[integer]:6 k[integer]:6
 COMMIT
 BEGIN
 table public.test_wal: INSERT: id[integer]:7 k[integer]:7
 table public.test_wal: INSERT: id[integer]:8 k[integer]:8
 COMMIT
(18 rows)

SELECT pg_drop_replication_slot('tde_slot');
 pg_drop_replication_slot 
--------------------------
 
(1 row)

DROP EXTENSION pg_tde;
